Detect integers 6644 v5 (#10122)
OISF/suricata: doc/userguide/rules/integer-keywords.rst

.. _rules-integer-keywords:

Integer Keywords
================

Many keywords will match on an integer value on the network traffic.
These are unsigned integers that can be 8, 16, 32 or 64 bits.

Simple example::

    bsize:integer value;

The integer value can be written as base-10 like ``100`` or as 
an hexadecimal value like ``0x64``.

> The most direct exemple is to match for equality, but there are

Author: fixing

different modes.

Comparison modes
----------------

Integers can be matched for
* Equality
* Inequality
* Greater than

> * Lesser than

Reviewer: This is one I've only recently learned that should be "Less than". My brain doesn't like it, but that's how it is '-'

Author: indeed

* Range
* Negated range
* Bitmask
* Negated Bitmask

> Comparison are strict by default.

Reviewer: nit:
```suggestion
Comparisons are strict by default.
```

Author: Plural indeed

> That means range between 1 and 4, will match 2 and 3, but not 1 neither 4.

Reviewer: nit:
```suggestion
That means a range between 1 and 4 will match 2 and 3, but not 1 or 4.
```

Author: neither nor ?

Reviewer: Used the suggestion given by grammarly, but to me neither nor also works.

Examples::

    bsize:integer value; # equality
    bsize:=integer value; # equality
    bsize:!integer value; # inequality
    bsize:!=integer value; # inequality
    bsize:>integer value; # greater than
    bsize:>=integer value; # greater than or equal
    bsize:<integer value; # lesser than
    bsize:<=integer value; # lesser than or equal
    bsize:integer value1-integer value2; # range between value1 and value2
    bsize:!integer value1-integer value2; # negated range between value1 and value2
    bsize:&mask=value; # bitmask mask is compared to value for equality
    bsize:&mask!=value; # bitmask mask is compared to value for inequality

Enumerations
------------

> Some integers on the wire represent an enumeration, that is some values

Reviewer: nit:
```suggestion
Some integers on the wire represent an enumeration, that is, some values
```

Author: ok

have a string/meaning associated to it.
Rules can be written using one of these strings to check for equality.
This is meant to make rules more human-readable and equivalent for matching.

Examples::

    websocker.opcode:text;

>     websocker.opcode:1; # behaves the same

Reviewer: Oh, is this something already implemented, or a new addition? 
@hadiqaalamdar this may mean that I was wrong with my previous review for your SV test, I'm sorry for that, I'll look into it further.

Author: This is one addition of this PR (one of the commits)

Reviewer: Ah, thanks!! Will be something nice to add to your SV tests if it gets merged soon, then, Hadiqa! :)

Bitmasks
--------

Some integers on the wire represent multiple bits.
Some of these bits have a string/meaning associated to it.

> Rules can be written using a list (comma-seperated) of these strings,

Reviewer: typo:
```suggestion
Rules can be written using a list (comma-separated) of these strings,
```

Author: indeed
